cas-client-core/src/main/java/org/jasig/cas/client/util/ThreadLocalXPathExpression.java

```diff
@@ -0,0 +1,87 @@
+package org.jasig.cas.client.util;
+
+import org.w3c.dom.NodeList;
+import org.xml.sax.InputSource;
+
+import javax.xml.namespace.NamespaceContext;
+import javax.xml.namespace.QName;
+import javax.xml.xpath.*;
+
+/**
+ * Thread local XPath expression.
+ *
+ * @author Marvin S. Addison
+ * @since 3.3.4
+ */
+public class ThreadLocalXPathExpression extends ThreadLocal<XPathExpression> implements XPathExpression {
```

Comment:
Fair enough. Personally, I feel that thread locals are helpful and justified in some cases. They are admittedly dangerous in library code, but the 99.9% use case here is integration with a Java Web application where thread locals abound anyway. I made this change in a focused commit called "sensible optimizations," but it's fair to say they're premature optimizations. On balance I don't see any harm given how the library is typically used.